### PR TITLE
Move NuGet SDK resolver to its own assembly

### PR DESCRIPTION
**Pull request by jeffkl**
Having a reference to NuGet but not shipping NuGet breaks NGEN and is a no-go.  This change moves it to its own assembly while leaving in all of the optimizations where NuGet assemblies are only loaded if needed.

1. New assembly
2. Moved resources
3. Updated SWR file, confirmed VSIX package layout
4. Updated bootstrap for local dev builds

### CODE REVIEW

**Comment by AndyGerlicher** [comment]
@livarcocc @nguerrera might want to take a look at the nuget packages. We need to make sure this will be compatible with the CLI.

File: build/NuGetPackages/Microsoft.Build.Localization.nuspec [hunk 1]
```diff
@@ -79,5 +79,21 @@
     <file src="$outputPath$tr\Microsoft.Build.Utilities.Core.resources.dll" target="lib\netstandard1.3\tr\Microsoft.Build.Utilities.Core.resources.dll" />
     <file src="$outputPath$zh-Hans\Microsoft.Build.Utilities.Core.resources.dll" target="lib\netstandard1.3\zh-Hans\Microsoft.Build.Utilities.Core.resources.dll" />
     <file src="$outputPath$zh-Hant\Microsoft.Build.Utilities.Core.resources.dll" target="lib\netstandard1.3\zh-Hant\Microsoft.Build.Utilities.Core.resources.dll" />
+
+    <!-- NuGet.MSBuildSdkResolver.dll -->
+    <file src="$outputPath$cs\NuGet.MSBuildSdkResolver.resources.dll" target="contentFiles\any\netstandard1.3\SdkResolvers\NuGet.MSBuildSdkResolver\cs\NuGet.MSBuildSdkResolver.resources.dll"/>
```
Comment (cdmihai) [inline build/NuGetPackages/Microsoft.Build.Localization.nuspec]: Put them in lib, like the rest. As far as I remember from my experiments, if you put them in content files, they don't show up in bin when the package is indirectly referenced.

File: build/NuGetPackages/Microsoft.Build.Runtime.nuspec [hunk 2]
```diff
@@ -90,5 +91,6 @@
     <file src="$outputPathNetCore$Microsoft.VisualStudioVersion.v11.Common.props" target="contentFiles\any\netcoreapp1.0\" buildAction="None" copyToOutput="true"/>
     <file src="$outputPathNetCore$Microsoft.VisualStudioVersion.v12.Common.props" target="contentFiles\any\netcoreapp1.0\" buildAction="None" copyToOutput="true"/>
     <file src="$outputPathNetCore$Microsoft.VisualStudioVersion.v14.Common.props" target="contentFiles\any\netcoreapp1.0\" buildAction="None" copyToOutput="true"/>
+    <file src="$outputPathNetCore$NuGet.MSBuildSdkResolver.dll" target="contentFiles\any\netcoreapp1.0\SdkResolvers\NuGet.MSBuildSdkResolver\" buildAction="None" copyToOutput="true"/>
```
Comment (cdmihai) [inline build/NuGetPackages/Microsoft.Build.Runtime.nuspec]: Putting them in this package means that if people reference only the Microsoft.Build package to evaluate projects, they won't get this resolver. Is this intended?

File: build/NuGetPackages/Microsoft.Build.Localization.nuspec (hunk 1 above)
Comment (nguerrera) [reply build/NuGetPackages/Microsoft.Build.Localization.nuspec]: I think it is OK as is. What I recall is that they were copied correctly from content files, but not used do to content not going in .deps.json file. But, in this case, the resolver will be loaded dynamically and the msbuild CoreCLR loader will find the satellites from relative path just like tasks.

File: setup/files.swr [hunk 3]
```diff
@@ -53,6 +53,38 @@ folder InstallDir:\MSBuild\15.0\Bin\MSBuild
...
+folder InstallDir:\MSBuild\15.0\Bin\SdkResolvers\NuGet.MSBuildSdkResolver\es
+  file source=$(X86BinPath)es\NuGet.MSBuildSdkResolver.resources.dll
+folder InstallDir:\MSBuild\15.0\Bin\SdkResolvers\NuGet.MSBuildSdkResolver\fr
+  file source=$(X86BinPath)fr\NuGet.MSBuildSdkResolver.resources.dll
+folder InstallDir:\MSBuild\15.0\Bin\SdkResolvers\NuGet.MSBuildSdkResolver\it
+  file source=$(X86BinPath)it\NuGet.MSBuildSdkResolver.resources.dll
+folder InstallDir:\MSBuild\15.0\Bin\SdkResolvers\NuGet.MSBuildSdkResolver\ja
+  file source=$(X86BinPath)ja\NuGet.MSBuildSdkResolver.resources.dll
+folder InstallDir:\MSBuild\15.0\Bin\SdkResolvers\NuGet.MSBuildSdkResolver\ko
+  file source=$(X86BinPath)ko\NuGet.MSBuildSdkResolver.resources.dll
```
Comment (cdmihai) [inline setup/files.swr]: I think you have to address the x64 bin dir as well. One approach would be to just update the lines below than extract the nuget ones separately.

File: setup/MsBuild.Engine.Corext/MsBuild.Engine.Corext.nuspec [hunk 4]
```diff
@@ -16,6 +16,8 @@
     <file src="$X86BinPath$/Microsoft.Common.props" target="\Extensions\15.0" />
     <file src="$X86BinPath$/Microsoft.VisualStudioVersion.v15.Common.props" target="\Extensions\15.0" />
 
+    <file src="$X86BinPath$/NuGet.MSBuildSdkResolver.dll" target="v15.0/bin/SdkResolvers/NuGet.MSBuildSdkResolver" />
```
Comment (cdmihai) [inline setup/MsBuild.Engine.Corext/MsBuild.Engine.Corext.nuspec]: what about x64?

File: build/NuGetPackages/Microsoft.Build.Localization.nuspec (hunk 1 above)
Comment (cdmihai) [reply build/NuGetPackages/Microsoft.Build.Localization.nuspec]: @nguerrera has a better memory than me :)

File: src/NuGetSdkResolver/NuGetSdkResolverBase.cs [hunk 5]
```diff
@@ -76,8 +99,7 @@ static NuGetSdkResolverBase()
             AssemblyName assemblyName)
         {
 #endif
-            // Only load NuGet.* and Newtonsoft.Json assemblies if found
-            if (assemblyName.Name.StartsWith("NuGet.", StringComparison.OrdinalIgnoreCase) || assemblyName.Name.Equals("Newtonsoft.Json", StringComparison.OrdinalIgnoreCase))
+            if (NuGetAssemblies.Contains(assemblyName.Name))
```
Comment (cdmihai) [inline src/NuGetSdkResolver/NuGetSdkResolverBase.cs]: Is it worth using an ordinal case ignoring comparer?

Comment (jeffkl) [reply src/NuGetSdkResolver/NuGetSdkResolverBase.cs]: Good point, I'll add it to the `HashSet<T>`

**Comment by cdmihai** [comment]
Another thing might be to update the microbuild checker with the new dlls, so we can test they're there and signed: https://github.com/Microsoft/msbuild/blob/master/build/MicrobuildTest.ps1#L36-L47

I can also do this later on next week when I'll update the checker for our new .net sdk build outputs.

File: build/NuGetPackages/Microsoft.Build.Framework.nuspec [hunk 6]
```diff
@@ -17,13 +17,13 @@
...
         <dependency id="System.Threading.Tasks.Parallel" version="4.0.1" />
         <dependency id="System.Runtime.Extensions" version="4.1.0" />
+        <dependency id="System.Resources.ResourceManager" version="4.0.1" />
         <dependency id="System.Collections" version="4.0.11" />
         <dependency id="System.Diagnostics.Debug" version="4.0.11" />
         <dependency id="System.Diagnostics.Process" version="4.1.0" />
         <dependency id="System.Globalization" version="4.0.11" />
         <dependency id="System.Linq" version="4.1.0" />
         <dependency id="System.ObjectModel" version="4.0.12" />
-        <dependency id="System.Resources.ResourceManager" version="4.0.1" />
```
Comment (cdmihai) [inline build/NuGetPackages/Microsoft.Build.Framework.nuspec]: Undo this change. For some reason the nuspec reference generator chooses a different ordering on the CI machines than on our local machines. :(

File: setup/MsBuild.Engine.Corext/MsBuild.Engine.Corext.nuspec (hunk 4 above)
Comment (AndyGerlicher) [reply setup/MsBuild.Engine.Corext/MsBuild.Engine.Corext.nuspec]: Resolvers only go in the 32 bit folder:
https://github.com/Microsoft/msbuild/blob/a955af860196952ba54a831cc2e57a91642d235a/src/Build/BackEnd/Components/SdkResolution/SdkResolverLoader.cs#L39-L40

File: build/NuGetPackages/Microsoft.Build.Framework.nuspec (hunk 6 above)
Comment (jeffkl) [reply build/NuGetPackages/Microsoft.Build.Framework.nuspec]: Oops this was committed by accident

File: build/NuGetPackages/Microsoft.Build.Runtime.nuspec (hunk 2 above)
Comment (jeffkl) [reply build/NuGetPackages/Microsoft.Build.Runtime.nuspec]: Yes, evaluation can only be done by referencing `Microsoft.Build.Runtime`

File: setup/files.swr (hunk 3 above)
Comment (jeffkl) [reply setup/files.swr]: The contents of the VSIX look correct

---
In reply to: [161095569](https://github.com/Microsoft/msbuild/pull/2855#discussion_r161095569) [](ancestors = 161095569)

File: src/Build/Resources/Strings.resx [hunk 7]
```diff
@@ -1208,21 +1208,9 @@
     <comment>{StrBegin="MSB4242: "}</comment>
   </data>
   <data name="CouldNotRunNuGetSdkResolver" UESanitized="false" Visibility="Public">
-    <value>MSB4243: The NuGet-based SDK resolver failed to run because NuGet assemblies could not be located.  Check your installation of MSBuild or set the environment variable "{0}" to the folder that contains the required NuGet assemblies. {1}</value>
+    <value>MSB4243: The NuGet-based SDK resolver failed to run because NuGet assemblies could not be located.  Check your installation of MSBuild or set the environment variable "MSBUILD_NUGET_PATH" to the folder that contains the required NuGet assemblies. {0}</value>
```
Comment (rainersigwald) [inline src/Build/Resources/Strings.resx]: Why not parameterized? Nice to have it in a single place.

File: src/NuGetSdkResolver/Properties/AssemblyInfo.cs [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+using System;
+using System.Reflection;
+using System.Resources;
+#if FEATURE_SECURITY_PERMISSIONS
+using System.Security.Permissions;
+#endif
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+
+#if FEATURE_SECURITY_PERMISSIONS
```
Comment (rainersigwald) [inline src/NuGetSdkResolver/Properties/AssemblyInfo.cs]: Does this assembly need all this junk? It seems excessive.

File: src/Build/BackEnd/Components/SdkResolution/SdkResolverService.cs [hunk 9]
```diff
@@ -123,14 +125,14 @@ public SdkResult GetSdkResult(int submissionId, SdkReference sdk, LoggingContext
                 {
                     result = (SdkResult) sdkResolver.Resolve(sdk, context, resultFactory);
                 }
-                catch (FileNotFoundException e) when (sdkResolver is NuGetSdkResolver)
+                catch (FileNotFoundException e) when (sdkResolver.GetType().GetTypeInfo().Name.Equals("NuGetSdkResolver", StringComparison.Ordinal))
```
[the remark this answers is not in the document]
Comment (jeffkl) [reply src/Build/BackEnd/Components/SdkResolution/SdkResolverService.cs]: yeah sorry it was the only option to keep the friendlier message.  Perhaps I should submit this line in a separate review as dotnet-bot so I don't get labeled as a traitor?

File: src/NuGetSdkResolver/Resources/AssemblyResources.cs [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+using System.Resources;
+using System.Reflection;
+using System.Globalization;
+
+namespace Microsoft.Build.Shared
+{
+    /// <summary>
+    /// This class provides access to the assembly's resources.
+    /// </summary>
+    internal static class AssemblyResources
```
Comment (rainersigwald) [inline src/NuGetSdkResolver/Resources/AssemblyResources.cs]: Can this be simplified for this assembly?

File: src/NuGetSdkResolver/NuGetSdkResolver.cs [hunk 11]
```diff
@@ -12,27 +12,33 @@
...
+    public sealed class NuGetSdkResolver : NuGetSdkResolverBase
     {
         public override string Name => nameof(NuGetSdkResolver);
 
         public override int Priority => 2500;
 
         public override SdkResultBase Resolve(SdkReference sdk, SdkResolverContextBase context, SdkResultFactoryBase factory)
         {
+            // Escape hatch to disable this resolver
+            if (Traits.Instance.EscapeHatches.DisableNuGetSdkResolver)
```
[the remark this answers is not in the document]
Comment (jeffkl) [reply src/NuGetSdkResolver/NuGetSdkResolver.cs]: Yes as long as this class has no direct dependency on NuGet, then this escape hatch can turn it off

File: src/NuGetSdkResolver/Resources/Strings.resx [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+  <data name="FailedToParseGlobalJson" UESanitized="false" Visibility="Public">
+    <value>Failed to parse "{0}". {1}</value>
+    <comment></comment>
+  </data>
+  <data name="CouldNotFindInstalledPackage" UESanitized="false" Visibility="Public">
+    <value>Failed to resolve SDK '{0}'. A package was successfully restored but the could not located.</value>
+    <comment></comment>
+  </data>
+  <data name="PackageWasNotInstalled" UESanitized="false" Visibility="Public">
+    <value>Failed to resolve SDK '{0}'. Package restore was successfully but a package with the ID of "{1}" was not installed.</value>
```
Comment (rainersigwald) [inline src/NuGetSdkResolver/Resources/Strings.resx]: "was successful"

File: src/Build/Resources/Strings.resx (hunk 7 above)
Comment (jeffkl) [reply src/Build/Resources/Strings.resx]: I've moved the environment variable name to MSBuildConstants and brought this back

File: src/Build/BackEnd/Components/SdkResolution/SdkResolverService.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (rainersigwald) [reply src/Build/BackEnd/Components/SdkResolution/SdkResolverService.cs]: You're going to make me institute an all-submissions-must-be-pgp-signed policy, aren't you :)

File: src/NuGetSdkResolver/Resources/AssemblyResources.cs (hunk 10 above)
Comment (jeffkl) [reply src/NuGetSdkResolver/Resources/AssemblyResources.cs]: Done

**Pull request merged**